platform/ios/src/MGLMapView.mm

```diff
@@ -5340,9 +5429,16 @@ - (void)cameraWillChangeAnimated:(BOOL)animated {
         }
     }
 
-    if ( ! [self isSuppressingChangeDelimiters] && [self.delegate respondsToSelector:@selector(mapView:regionWillChangeAnimated:)])
+    if ( ! [self isSuppressingChangeDelimiters] )
     {
-        [self.delegate mapView:self regionWillChangeAnimated:animated];
+        if ([self.delegate respondsToSelector:@selector(mapView:regionWillChangeAnimated:reason:)])
+        {
+            [self.delegate mapView:self regionWillChangeAnimated:animated reason:self.cameraChangeReason];
+        }
+        else if ([self.delegate respondsToSelector:@selector(mapView:regionWillChangeAnimated:)])
```

Comment:
If the delegate happens to implement both variations of the method, MGLMapView should call both in succession. We can document the order in which these methods are called.

I think we shouldn’t ignore an existing method implementation when the more specific variation is implemented. Consider [a very long delegate implementation](https://github.com/mapbox/mapbox-navigation-ios/blob/master/MapboxNavigation/RouteMapViewController.swift), in which it’s difficult to keep track of which optional delegate methods have been implemented. I’m oblivious to the fact that `-mapView:regionWillChangeAnimated:` has already been implemented in this class, and I notice that `-mapView:regionWillChangeAnimated:` has been deprecated in favor of `-mapView:regionWillChangeAnimated:region:`. So I implement `-mapView:regionWillChangeAnimated:region:`, and eventually we discover that some unrelated functionality has broken because `-mapView:regionWillChangeAnimated:` has become dead code.

Reply:
I'm not in favour of calling both, I think it adds some confusion for the developer. I've added some documentation saying that the previous method will be deprecated in future.

Perhaps it's better to the older methods as deprecated immediately? Thoughts @akitchen ?

What are the conventions for deprecation? Do we comment first, then in a later release add the deprecation attribute, then in a further major release remove them?

Reply:
Normally, we’ll deprecate the symbol in one release (potentially a patch release) then mark it unavailable in a later release. But we haven’t had a major version in a couple years, so we haven’t had the opportunity to mark anything unavailable.

The main thing going for the older methods at this point is familiarity/compatibility with MapKit, but this PR departs from MapKit’s APIs anyways. If we don’t deprecate the older methods, then we should either call both methods (and document the entire lifecycle of delegate methods in an ordered list) or at least issue a one-time warning in the console if both methods are implemented. Otherwise, it’s very likely that the confusion described in https://github.com/mapbox/mapbox-gl-native/pull/11151#discussion_r167128929 will occur (probably first in the navigation SDK). The fallout will be difficult to debug, since it’s always difficult to debug why a method _isn’t_ getting called.